src/neuroconv/utils/path_expansion.py

```diff
@@ -0,0 +1,100 @@
+import abc
+import os
+from collections import defaultdict
+from glob import glob
+from typing import Dict, List, Union
+
+from parse import parse
+
+from .types import FilePathType, FolderPathType
+
+
+def _ddict():
+    """
+    Create a defaultdict of defaultdicts
+
+    This allows you to easily nest hierarchical dictionaries. For example, this syntax
+    >>> a = dict(b=dict(c=dict(d=5)))
+
+    becomes
+    >>> a = _ddict()["b"]["c"]["d"] = 5
+
+    It becomes particularly useful when modifying an existing hierarchical dictionary,
+    because the next level is only created if it does not already exist.
+
+    """
+    return defaultdict(_ddict)
+
+
+def _unddict(d):
+    """Turn a ddict into a normal dictionary"""
+    return {key: _unddict(value) for key, value in d.items()} if isinstance(d, defaultdict) else d
+
+
+class AbstractPathExpander(abc.ABC):
+    def extract_metadata(self, folder, format_: str):
+        for filepath in self.list_directory(folder):
+            result = parse(format_, filepath)
```

Comment:
After playing around with this both in Python and GUIDE, I think it would be a better user experience if we constrain our f-string language to always use "/" to indicate file path separators independent of the OS; this is easy to correct in the format passed into `parse` like so

```suggestion
class AbstractPathExpander(abc.ABC):
    def extract_metadata(self, folder, format_: str):
        format_ = format_.replace("/", os.sep)  # our f-string uses '/' to communicate os-independent separators
        for filepath in self.list_directory(folder):
            result = parse(format_, filepath)
```